Special bug

lsgunth opened the pull request:
A couple more fixes starting with a bug in the special disk handling for the journal.

lsgunth committed:
md: Don't try to figure out disk type from sysfs in _get_next_disk()

get_special_disk() is called in setup() before the array is setup, so
trying to read from sysfs to get the disk (or size) for the journal
does not work.

Change it to store the disk_type and disk_size and use that for
creating extra ram or loop disks.

lsgunth committed:
md: Drop journal argument

Now that the journal argument doesn't specify the device for the
journal it is redundant with policy. Remove it.

lsgunth committed:
test_all: Move test2 test down

The original intention was for fio to run in the test2 environment,
not one recently degraded. That's why there was originally a
second test2 run. So just move it down.

lsgunth committed:
md: Create loop disks in start()

If the loop disks are created before the previous array is stopped,
the creation will fail as they are already in use.

Move the creation into start().

Shouldn't be doing anything that has side effects in __init__ anyway.

iomartin reviewed (approved):
Looks good to me

lsgunth committed:
test3: Don't output to stderr

Now that --verify is included in test_all, we don't want the extra
noise on stderr.

dlsloan reviewed (dismissed):
Looks good to me!

lsgunth committed:
test3: Fix disk zeroing in test3 --verify

Move the disk zeroing code to md so it can handle it appropriately
and zero new grown disks and work correctly before setup() is called.

lsgunth commented:
Added one more fixup

dlsloan commented on md.py:
```diff
@@ -197,6 +197,8 @@ def __init__(self, md="md0", level=5, devs=None, ndisks=None,
             if size is None:
                 raise MDInvalidArgumentError("Must specify size with loop_disks")
+            #Placeholder that will be recreated in setup()
+            self.devs = [None] * ndisks
```
test3 --verify actually needs the disks to be created for zeroing

lsgunth committed:
test_all: Implement quick test

Implement a --quick option to test_all. This is primarily for ensuring
the test suite works when md.py is changed.

lsgunth commented:
Ok revamped the disk zeroing, it should be much improved.

dlsloan commented on md.py:
```diff
@@ -134,6 +134,13 @@ class MDInvalidArgumentError(Exception):
     pass
 
 class MDInstance:
+    # Raid superblock v1.2 consumes (256 + 2 * ndisks) bytes at a 4k offset
```
Comment needs an update with the size change

lsgunth replied:
Fixed

dlsloan commented on md.py:
```diff
@@ -281,11 +293,30 @@ def get_num_disks(self):
+
+        count = size_to_zero + self.MAX_SUPERBLOCK_SZ
+        count = (count + (1 << 20) - 1) >> 20
+        self._size_to_zero = count
```
Should we add a dev size limit here in case we have very small ram disks or large zeroing ranges? Or id dd ok with count*bs > dev size?

lsgunth replied:
It'll fail, but it's probably good to fail. If the device isn't at least that large the test will fail in other ways.

dlsloan replied:
We only need to zero size_to_zero / ndata + data_offset, so for small disks/large regions we're limiting the test size to a fraction of the disk, Might be a good idea to be able to test the end of the array as well as the beginning.

dlsloan replied:
On second thought, we're using a spare disk as the veification scratch pad so end of disk tests will need special consideration anyway, this is good as is.

dlsloan commented on test_all:
```diff
@@ -27,6 +27,14 @@ def run_all_tests(env):
+    run_test(["./test3", "-d", "--quick"], env)
+    if not env or env.get("NO_RESHAPE") != "y":
+        run_test(["./test3", "-g", "--quick"], env)
+    run_test(["./test2"], env)
```
Big fan of this! I've basically been doing this manually for my testing
